packages/@eventual/core/src/entity/entity.ts

```diff
@@ -298,6 +312,46 @@ export function entity<
         }
       );
     },
+    index: (...args) => {
+      const [indexName, indexOptions] = args;
+
+      if (indices.some((i) => i.name === indexName)) {
+        throw new Error(`Index of name ${indexName} already exists on ${name}`);
+      }
+
+      const index: EntityIndex = {
+        kind: "EntityIndex",
+        key: computeKeyDefinition(
+          attributes,
+          ("partition" in indexOptions ? indexOptions.partition : undefined) ??
+            options.partition,
+          indexOptions.sort ?? options.sort
+        ),
+        name: indexName,
+        partition:
+          "partition" in indexOptions ? indexOptions.partition : undefined,
+        sort: indexOptions.sort,
+        entityName: name,
+        query: (...args) => {
+          return getEventualCallHook().registerEventualCall(
+            createEventualCall<EntityCall<"queryIndex">>(
+              EventualCallKind.EntityCall,
+              {
+                entityName: name,
+                indexName,
+                operation: "queryIndex",
+                params: args,
+              }
+            ),
+            () => getEntityHook().queryIndex(name, indexName, ...args)
+          );
+        },
```

Comment:
Indexes can be scanned too right?

Reply:
Yes, but we don't expose that api right now for entities.